crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs

```diff
@@ -0,0 +1,554 @@
+}
+
+impl RemoteSequencerConnection {
+    fn start(
+        known_global_tail: TailOffsetWatch,
+        connection: WeakConnection,
+    ) -> Result<Self, ShutdownError> {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        task_center().spawn(
+            TaskKind::Disposable,
+            "remote-sequencer-connection",
+            None,
+            Self::handle_appended_responses(known_global_tail, connection.clone(), rx),
+        )?;
+
+        Ok(Self {
+            inner: connection,
+            tx,
+        })
+    }
+
+    /// Send append message to remote sequencer.
+    ///
+    /// It's up to the caller to retry on [`NetworkError`]
+    pub async fn send(
+        &self,
+        rpc_router: &RpcRouter<Append>,
+        sequencer: GenerationalNodeId,
+        msg: Append,
+    ) -> Result<RpcToken<Appended>, NetworkSendError<Append>> {
+        let outgoing = Outgoing::new(sequencer, msg).assign_connection(self.inner.clone());
+
+        rpc_router
+            .send_on_connection(outgoing)
+            .await
+            .map_err(|err| NetworkSendError::new(err.original.into_body(), err.source))
+    }
+
+    pub fn resolve_on_appended(
+        &self,
+        permit: OwnedSemaphorePermit,
+        rpc_token: RpcToken<Appended>,
+        commit_resolver: LogletCommitResolver,
+    ) {
+        let inflight_append = RemoteInflightAppend {
+            rpc_token,
+            commit_resolver,
+            permit,
+        };
+
+        if let Err(err) = self.tx.send(inflight_append) {
+            // if we failed to push this to be processed by the connection reactor task
+            // then we need to notify the caller
+            err.0
+                .commit_resolver
+                .error(AppendError::retryable(NetworkError::ConnectionClosed));
+        }
+    }
+
+    /// Handle all [`Appended`] responses
+    ///
+    /// This task will run until the [`AppendStream`] is dropped. Once dropped
+    /// all pending commits will be resolved with an error. it's up to the enqueuer
+    /// to retry if needed.
+    async fn handle_appended_responses(
+        known_global_tail: TailOffsetWatch,
+        connection: WeakConnection,
+        mut rx: mpsc::UnboundedReceiver<RemoteInflightAppend>,
+    ) -> anyhow::Result<()> {
+        let mut closed = std::pin::pin!(connection.closed());
+
+        // handle all rpc tokens in a loop, this loop will
+        // break only with a break reason.
+        // this reason can be one of the following:
+        // - Shutdown error
+        // - Terminal AppendError
+        // - Connection closed
+        let reason = loop {
+            let inflight = match rx.recv().await {
+                Some(inflight) => inflight,
+                None => {
+                    break BreakReason::Shutdown(ShutdownError);
+                }
+            };
+
+            tokio::select! {
+                appended = inflight.rpc_token.recv() => {
```

Comment:
I assume that the response for an `Append` message will be sent back on the same connection we used for sending the request, right? Otherwise, this message might get lost w/o us noticing it which would result into being stuck here.

Reply:
Yes, that's correct. It's one of the requirements for this to work. The sequencer must reply on the same connection